dbms/src/Flash/Planner/Plans/PhysicalJoin.cpp

```diff
@@ -324,14 +323,58 @@ void PhysicalJoin::buildPipeline(PipelineBuilder & builder, Context & context, P
     builder.addPlanNode(join_probe);
 }
 
-void PhysicalJoin::finalize(const Names & parent_require)
+void PhysicalJoin::finalizeImpl(const Names & parent_require)
 {
-    // schema.size() >= parent_require.size()
     FinalizeHelper::checkSchemaContainsParentRequire(schema, parent_require);
+    join_ptr->finalize(parent_require);
+    auto required_input_columns = join_ptr->getRequiredColumns();
+    std::unordered_set<String> build_schema_sets;
+    for (const auto & name : build_side_prepare_actions->getSampleBlock().getNames())
+    {
+        build_schema_sets.insert(name);
+    }
+    std::unordered_set<String> probe_schema_sets;
+    for (const auto & name : probe_side_prepare_actions->getSampleBlock().getNames())
+    {
+        probe_schema_sets.insert(name);
+    }
+    Names build_required;
+    Names probe_required;
+    bool has_missed_column = false;
+    String missed_column_name;
+    for (const auto & name : required_input_columns)
+    {
+        if (build_schema_sets.find(name) != build_schema_sets.end())
+            build_required.push_back(name);
+        else if (probe_schema_sets.find(name) != probe_schema_sets.end())
+            probe_required.push_back(name);
+        else
+        {
+            has_missed_column = true;
+            missed_column_name = name;
+            break;
+        }
+    }
+    if (has_missed_column)
+        throw Exception("Meet unknown column: {}", missed_column_name);
+    build_side_prepare_actions->finalize(build_required);
+    auto child_required_columns = build_side_prepare_actions->getRequiredColumns();
+    if unlikely (child_required_columns.empty())
+    {
+        child_required_columns.push_back(ExpressionActions::getSmallestColumn(build()->getSchema()));
+    }
+    build()->finalize(child_required_columns);
+    probe_side_prepare_actions->finalize(probe_required);
+    child_required_columns = probe_side_prepare_actions->getRequiredColumns();
+    if unlikely (child_required_columns.empty())
+    {
+        child_required_columns.push_back(ExpressionActions::getSmallestColumn(probe()->getSchema()));
+    }
```

Comment:
Seems useless since this case is taken into account in ExpressionAction?
https://github.com/pingcap/tiflash/blob/adf85696270f864c0317526919b37fbdfe5a6417/dbms/src/Interpreters/ExpressionActions.cpp#L647-L653

Reply:
Yes, I've removed it.